File: LeetCode_DailyPractice/day14_linked_list/day14_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class MyLinkedList:
    def __init__(self):
        self.dummy_head = ListNode()
        self.size = 0

    def get(self, index: int) -> int:         # : int 是返回类型提示（type hint），表示 val 应该是一个整数。类型提示是可选的，但它们可以帮助开发者理解代码的预期输入类型。
        if index < 0 or index >= self.size:
            return -1
        
        current = self.dummy_head.next
        for i in range(index):
            current = current.next
            
        return current.val

    def addAtHead(self, val: int) -> None:     #  -> None 是返回类型提示，表示这个方法不返回任何值。
        self.dummy_head.next = ListNode(val, self.dummy_head.next)
        self.size += 1

    def addAtTail(self, val: int) -> None:
        current = self.dummy_head
        while current.next:
            current = current.next
        current.next = ListNode(val)
        self.size += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.size:
            return
        
        current = self.dummy_head
        for i in range(index):
            current = current.next
        current.next = ListNode(val, current.next)
        self.size += 1

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return
        
        current = self.dummy_head
        for i in range(index):
            current = current.next
        current.next = current.next.next
        self.size -= 1

    def printList(self) -> None:
        current = self.dummy_head
        
        while current.next:
            current = current.next
            print(current.val)

if __name__ == "__main__":
    List = MyLinkedList()
    List.addAtHead(7)
    List.addAtHead(2)
    List.addAtHead(1)
    List.addAtIndex(3,0)
    List.deleteAtIndex(2)
    List.addAtHead(6)
    List.addAtTail(4)
    List.get(4)
    List.addAtHead(4)
    List.addAtIndex(5,0)
    List.addAtHead(6)
    List.printList()

File: LeetCode_DailyPractice/day14_linked_list/day14_linked_list.cpp
// leetcode 707
#include<iostream>

using namespace std;

struct ListNode{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int val) : val(val), next(nullptr) {}
    ListNode(int val, ListNode *next) : val(val), next(next) {}

};

class MyLinkedList {
    private:
        int size;
        ListNode *dummyhead;

    public:
        // 类构造函数 
        MyLinkedList() {
            this->size = 0;
            this->dummyhead = new ListNode(0);       //虚拟头节点
        }
        
        // 根据下标获取数值(下标从0开始)
        int get(int index) {
            if(index < 0 || index > (size - 1))
                return -1;
        
            ListNode *cur = dummyhead -> next;       // 注意查询时cur指向的是当前节点，所以开始时从虚拟头节点后开始
            while(index){
                cur = cur -> next;
                index --;
            }
            return cur -> val;
        }

        // 添加新的头节点
        void addAtHead(int val) {
            ListNode *newnode = new ListNode(val);

            newnode -> next = dummyhead -> next;
            dummyhead -> next = newnode;

            size += 1;
        }
        
        // 添加新的尾节点 
        void addAtTail(int val) {
            ListNode *cur = dummyhead;

            while(cur -> next != nullptr)
                cur = cur -> next;
        
            ListNode *newnode = new ListNode(val);
            cur -> next = newnode;

            size += 1;
        }

        // 在指定位置添加新节点
        void addAtIndex(int index, int val) {
            if(index > size) return;                    // index大于size甚至已经超出链表的尾部（此处要谨慎debug）
            if(index < 0) index = 0;
            ListNode *cur = dummyhead;
            ListNode *newnode = new ListNode(val);
            
            // 循环量要代入一些极端值进行判断
            while(index){
                cur = cur -> next;
                index --;
            }
            newnode -> next = cur -> next;
            cur -> next = newnode;
            
            size += 1;
        }

        // 在指定位置删除节点
        void deleteAtIndex(int index) {
            if(index >= 0 && index < size){
                ListNode *cur = dummyhead;
            
                while(index){
                    cur = cur -> next;
                    index --;
                }

                ListNode *tmp = cur -> next;
                cur -> next = tmp -> next;
                // cur -> next = cur -> next -> next;
                delete tmp;
                tmp = nullptr;
                size -= 1;
            }
        }

        // 打印链表
        void printList(){
            ListNode *cur = dummyhead;
            while(cur -> next != nullptr){
                cur = cur -> next;
                cout << cur -> val << " ";
            }
        }
};

// 测试+debug
int main()
{
    MyLinkedList list;
    list.addAtHead(7);
    list.addAtHead(2);
    list.addAtHead(1);
    list.addAtIndex(3,0);
    list.deleteAtIndex(2);
    list.addAtHead(6);
    list.addAtTail(4);
    list.get(4);
    list.addAtHead(4);
    list.addAtIndex(5,0);
    list.addAtHead(6);
    list.printList();
    return 0;
}